### PR TITLE
Onnx support: modify unsqueeze function

### PR DESCRIPTION
**Pull request by lingxiao1989**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this patch)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If it is possible, please attach a screenshot; otherwise, remove this)

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala [hunk 1]
```diff
@@ -32,46 +32,51 @@ import scala.reflect.ClassTag
...
-
-    // the actual position; clearer error message for batchMode (if any)
-    val actualPos = dim + offsetDim
-    require(actualPos >= 1 && actualPos <= (inputDim + 1), s"Invalid position: $pos. " +
-      s"input:dim() is $input, input feature map dim (numInputDims) is $numInputDims.")
-
-    actualPos
+  private def getActualPosition(input: Tensor[_]) : Array[Int] = {
+    pos.foreach(pos => {
+      val dim = if (pos <= 0) {
```
Comment (LeicongLi) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: Please try to simplify the logic here.
Suggest get rid of If Else statement.
The Require statement enforces pos > 0.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala [hunk 2]
```diff
@@ -32,46 +32,51 @@ import scala.reflect.ClassTag
...
+        require(pos <= 0, s"invalid input: index starts from 1, " +
```
Comment (LeicongLi) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: The condition is opposite, we want pos > 0, right?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala [hunk 3]
```diff
@@ -32,46 +32,51 @@ import scala.reflect.ClassTag
...
+      // the actual position; clearer error message for batchMode (if any)
+      val actualPos = dim + offsetDim
+      require(actualPos >= 1 && actualPos <= (inputDim + 1), s"Invalid position: $pos. " +
+        s"input:dim() is $input, input feature map dim (numInputDims) is $numInputDims.")
+      actualPos
+    })
+    for {
+      i <- 1 to pos.length
+      j <- i to pos.length
+    } if (pos(j) > pos(i)) { pos(j) = pos(j) + 1}
```
Comment (LeicongLi) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: Please add comment/explanation to this code.
j <- i + 1 to pos.length would make it easier to understand?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala [hunk 4]
```diff
@@ -108,6 +113,12 @@ object Unsqueeze {
   def apply[@specialized(Float, Double) T: ClassTag](
     pos: Int,
     numInputDims: Int = Int.MinValue)(implicit ev: TensorNumeric[T]) : Unsqueeze[T] = {
-    new Unsqueeze[T](pos, numInputDims)
+    new Unsqueeze[T](Array(pos), numInputDims)
+  }
+
+  def apply[@specialized(Float, Double) T: ClassTag](
```
Comment (LeicongLi) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: Please add Python API for this.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala [hunk 5]
```diff
@@ -32,46 +32,51 @@ import scala.reflect.ClassTag
...
     if (input.getType() != output.getType()) {
       output = input.emptyInstance()
     }
-
-    output
+    actualPos.foreach(output
       .asInstanceOf[Tensor[NumericWildcard]]
-      .addSingletonDimension(input.asInstanceOf[Tensor[NumericWildcard]], actualPos)
-
+      .addSingletonDimension(input.asInstanceOf[Tensor[NumericWildcard]], _))
```
Comment (LeicongLi) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: Can we do the insertions at same time?
The for loop hurts performance.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala (hunk 3 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: Yes, ok

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala (hunk 1 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: the original function takes 0 as input.dim() + pos + 1. Are you suggesting we take this out and only take numbers >0 ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala (hunk 4 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: ok

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala (hunk 5 above)
Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: I think the loop takes more time than one insertion call, but just a linear increase. We will still use a loop to add multiple dimensions to a tensor, right?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala (hunk 1 above)
Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: How are you going to treat the input if it is 2, do you know it is starting from 0 or 1?

Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: the code assume the input start from 1, and there is a require to make sure the pos is no larger than the input size:
require(actualPos >= 1 && actualPos <= (inputDim + 1), s"Invalid position: $pos. " +
        s"input:dim() is $input, input feature map dim (numInputDims) is $numInputDims.")

Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Unsqueeze.scala]: I dont understand.
The If statement here guarantee pos <= 0, why there is a require statement after it?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 6]
```diff
@@ -1531,6 +1531,13 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
       numInputDims)
   }
 
+  def createMultipleDimUnsqueeze(pos: Array[Int],
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: I don't think we need 2 separate bridge method here, please consolidate it into one,  removing above one and use this one works for both.

Comment (lingxiao1989) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: Ok, I have change it into one method.

**Comment by lingxiao1989** [comment]
Jenkins:
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/915/

**Comment by lingxiao1989** [comment]
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/918/

**Pull request merged**